File: bin/rmLinker.py
import numpy, itertools
from Bio.Seq import Seq

##########################################################
# Functions
def locate(queries,read,numMismatches):
	read =numpy.array( map(str,read)  )
	minMismatches=numMismatches+1 # dummy initial value
	matchLength=0   # dummy initial value
	output  = [ 0, len(queries.itervalues().next()), 0, len(read), 'none','none'] # startQ, endQ, startR, endR, strand, Type('match','trim','none')
	for orientation in queries: # Iterates over forward & reverse complemented queries
		query = queries[orientation]
		if orientation == "f" or orientation == "fc":
			strand = "+"
		elif orientation == "r" or orientation == "rc":
			strand = "-"
		for i in range( 0, len(read) ):
			startR = i
			endR   = min( len(read) , i+len(query) )
			startQ = 0
			endQ   = min( len(query), startQ	+(endR-startR) )
			# Allow only 2 mistmatches
			mismatches = sum(query[startQ:endQ]!=read[startR:endR])			   
			if (mismatches <= numMismatches) and (mismatches < minMismatches) and (endR-startR)>matchLength:
				minMismatches = mismatches
				matchLength   = endR - startR
				# Require at least 15 nt to call a match
				if matchLength>=15:
					output = [ startQ, endQ, startR, endR, strand, 'match' ]
				else: # If match happens in the last 15 nt assumed is just by chance and use it only for trimming
					output = [ startQ, endQ, startR, endR, strand, 'trim' ]				
	return output

def printMates(queries,mate1,mate2,numMismatches, detailsFile, mate1File, mate2File):

	read_id_f, read_seq_f, read_id2_f, read_qual_f = mate1
	read_id_f, read_seq_s, read_id2_s, read_qual_s = mate2
	startQ_f,endQ_f,startR_f,endR_f,strand_f, Type_f = locate(queries,read_seq_f,numMismatches)
	startQ_s,endQ_s,startR_s,endR_s,strand_s, Type_s = locate(queries,read_seq_s,numMismatches)
	# one or two ends with linker sequence at the beginning
	if startR_f<=15 or startR_s<=15:
		classification="ambiguous"
	# Coherent mates
	elif Type_f=="match" and Type_s=="match" and strand_f!=strand_s:
		classification="coherent_mates"
	# Un-coherent mates
	elif Type_f=="match" and Type_s=="match" and strand_f==strand_s:
		classification="uncoherent_mates"
	# only one matches
	elif (Type_f=="match" and Type_s!="match") or (Type_f!="match" and Type_s=="match"):
		classification="one_mate_linker"
	# none matches
	elif Type_f!="match" and Type_s!="match":   
		classification="no_linker"

	# Chop sequences and qualities
	read_seq_f, read_qual_f = read_seq_f[:startR_f], read_qual_f[:startR_f]
	read_seq_s, read_qual_s = read_seq_s[:startR_s], read_qual_s[:startR_s]   
	# Print details 
	details  = [ read_id_f ,startR_f,endR_f,strand_f,Type_f ]
	details +=		  [ startR_s,endR_s,strand_s,Type_s, classification ]
	print >> detailsFile, '\t'.join(map(str,details))
	# Print only non ambiguos fastq mates and attach classification to read_id
	if classification != "ambiguous":
		read_id = read_id_f + "__" + classification
		read_id2 = read_id2_f + "__" + classification
		print >> mate1File, '\n'.join(map(str,[read_id, read_seq_f, read_id2, read_qual_f]))
		print >> mate2File, '\n'.join(map(str,[read_id, read_seq_s, read_id2, read_qual_s]))


###########################################################################3

query="CTAGTAGCCCATGCAATGCGAGGA"
fastqFile1="/data2/sysbio/UCSD-sequencing/2014-3-26-Tri/CCGGRm_dupPE_RNA-RNA-03-26-2014_R1.fastq"
fastqFile2="/data2/sysbio/UCSD-sequencing/2014-3-26-Tri/CCGGRm_dupPE_RNA-RNA-03-26-2014_R2.fastq"
numMismatches=2

detailsFile=open("matesDetails.txt",'w')
mate1File  =open("mate1.fastq",'w')
mate2File  =open("mate2.fastq",'w')

# Excecution
# Define all possible orientations of the query
queries={} # stores forward and reverse complemented query
forward  = map(str,query)
forward_complement = map(str,Seq(query).complement())
reverse = map(str, str( query[::-1] ) )
reverse_complement = map(str, str( Seq(query).reverse_complement() ) )
queries["f"]=numpy.array( forward  )
queries["r"]=numpy.array( reverse )
queries["fc"]=numpy.array( forward_complement  )
queries["rc"]=numpy.array( reverse_complement )

# Loop over fastq reads
fastq1=open(fastqFile1,'r')
fastq2=open(fastqFile2,'r')
# Read mates
mate1=['dummy','dummy','dummy','dummy']
count=0
while mate1[0] != '':
	count += 1
	if count%100==0: print count
	mate1 = [fastq1.readline().strip() for i in range(4)]
	mate2 = [fastq2.readline().strip() for i in range(4)]
	# On mates, chop/trim linker sequence, classify, and print them
	printMates(queries,mate1,mate2,numMismatches,detailsFile,mate1File,mate2File)

fastq1.close()
fastq2.close()
detailsFile.close()
mate1File.close()
mate2File.close()

File: bin/rmLinker.cpp
#include <iostream>
#include <string>
#include <vector>
#include <iostream> // input and output files
#include <fstream>  
#include <map>
#include <sstream>  // for to_string function
#include <unistd.h> // get options
#include <getopt.h> // get long options
#include <stdlib.h> // atoi
using namespace std;
/** Remove linker sequence from paired-end mates */

// Functions ######################################################################
//#################################################################################
void print_usage() 
{
	printf("================================================\n");
	printf("Remove linker sequence from paired-end reads.\n");
    printf("Usage: rmLinker -1 fastq1 -2 fastq2 -l forward_linker_sequence [OPTIONS]\n");
    printf("Options:\n");
    printf("\t-p --prefix\tPrefix of output files (two fastq, and one txt files). Default: 'result'.\n");
    printf("\t-m --mismatches\tNumber of mismatches allowed. Default: 2.\n");
    printf("\t-f --direction1\tComma separated list of allowed linker directions on the forward mate. Default: 'fwd,fwdC,rev,revC'.\n");
    printf("\t-r --direction2\tComma separated list of allowed linker directions on the reverse mate. Default: 'fwd,fwdC,rev,revC'.\n");
}
template <typename T>
string to_string(T value)
{
  //create an output string stream
  ostringstream os ;
  //throw the value into the string stream
  os << value ;
  //convert the string stream into a string and return
  return os.str() ;
}
//#################################################################################
int countMismatches(string const & read, string const & query)
{
    int localScore = 0;
    for (unsigned i = 0; i < query.size(); ++i)
        if (read[i] != query[i])
            ++localScore;
    return localScore;
}
//#################################################################################
struct alignment {
	int start;
	int end;
	string strand;
	string type;
};
//#################################################################################
alignment computeAlignment(map<string,string> & queries, string const & read, int const & numMismatches) 
{
//	string read = mate["seq"];
    int mismatches, minMismatches;
	int matchLength=0;
	int startR, endR, startQ, endQ;
	alignment output;
	output.start = read.size(); output.end=read.size(); output.strand="none"; output.type="none";
	string query, strand;

	minMismatches = numMismatches + 1; 
	matchLength   = 0;
	// Loop over queries
    map<string, string>::iterator iter;
    for( iter=queries.begin(); iter!=queries.end(); ++iter ){
		strand = iter->first; 
		query  = iter->second;
		// Iterate over read length
		for (unsigned i = 0; i < read.size() ; ++i){
			startR = i; // start of read
			endR   = min( read.size(), i + query.size() ); // end of read
			startQ = 0; // start of query
			endQ   = min<unsigned int>( query.size(), startQ +(endR-startR) ); // end of query
			mismatches = countMismatches( read.substr(startR, endR-startR ), query.substr(startQ,endQ-startQ) );

			if( (mismatches <= numMismatches) and (mismatches < minMismatches) and ((endR-startR)>matchLength) ){
				minMismatches = mismatches;
				matchLength   = endR - startR;
				//Requiere at least 15 nt to call a match
				if( matchLength >=15 ){
					output.start  = startR; output.end = endR; 
					output.strand = strand; output.type = "match";
				}else if (mismatches==0){ // If match happens in beyond the last 15nt, assume is only by chance and use it for trimming. Don't tolerate mismatches here.
					output.start  = startR; output.end = endR; 
					output.strand = strand; output.type = "trim";
	}	}	}	}
    return output;
}
// ##########################################################################################
struct readID{
	string read_idS; // sequence read id
	string read_idQ; // quality read id
};
readID getRead_ids(string const & read_id_original,string const & classification){
	readID ids;
	string prefix, suffix;
	stringstream read_id_stream(read_id_original.substr(1));
	getline(read_id_stream,prefix,' ');
	getline(read_id_stream,suffix,' ');
	ids.read_idS = "@" + prefix + "_type_" + classification + " " + suffix; 
	ids.read_idQ = "+" + prefix + "_type_" + classification + " " + suffix;
	return ids;
}
// ##########################################################################################
int printMates(map<string,string> & queries1, map<string,string> & queries2, map<string,string> & mate1, map<string,string> & mate2, int const & numMismatches, ofstream & detailsFile, ofstream & mate1File, ofstream & mate2File)
{
	// Compute alignments
	alignment align1 = computeAlignment( queries1, mate1["seq"], numMismatches);
	alignment align2 = computeAlignment( queries2, mate2["seq"], numMismatches);
	// Classify mates
	string classification = align1.strand+"_"+align2.strand;
	// Chop fastq sequences and qualities
	string seq1  = mate1["seq" ].substr( 0, align1.start );
	string qual1 = mate1["qual"].substr( 0, align1.start );
	string seq2  = mate2["seq" ].substr( 0, align2.start );
	string qual2 = mate2["qual"].substr( 0, align2.start );
	// Print details
	detailsFile << mate1["id"]+"\t"+to_string(align1.start)+"\t"+to_string(align1.end)+"\t"+align1.strand+"\t"+align1.type+"\t"+to_string(align2.start)+"\t"+to_string(align2.end)+"\t"+align2.strand+"\t"+align2.type+"\n";
	// If reads are to short, or linker are not oriented properly, don't print them.
	if ( (align1.start>=15) and (align2.start>=15)  ){
		readID id1 = getRead_ids( mate1["id"], classification);
		readID id2 = getRead_ids( mate2["id"], classification);
		mate1File << id1.read_idS +"\n"+seq1+"\n"+id1.read_idQ+"\n"+qual1+"\n";
		mate2File << id2.read_idS +"\n"+seq2+"\n"+id2.read_idQ+"\n"+qual2+"\n";
	}
	return 0;
}
// ##########################################################################################
string getComplement( string const & forward )
{
	string complement;
    for( string::size_type i = 0; i < forward.size(); ++i ) {
        if      ( (forward[i] == 'A') or (forward[i] == 'a') ) complement += 'T';
        else if ( (forward[i] == 'C') or (forward[i] == 'c') ) complement += 'G';
        else if ( (forward[i] == 'G') or (forward[i] == 'g') ) complement += 'C';
        else if ( (forward[i] == 'T') or (forward[i] == 't') ) complement += 'A';
    }	
	return complement;
}
// ##########################################################################################
string getReverse( string const & forward )
{
    string reversed="";
    for( unsigned i=0; i<forward.length(); ++i) {
        reversed = forward[i]+reversed ;
    }
    return reversed;
}
// Execution ################################################################################
//###########################################################################################
int main(int argc, char *argv[])
{
	// Get arguments ########################################################################
	string fastq1="none", fastq2="none", linker="none";
	string direction1="fwd,fwdC,rev,revC";
	string direction2="fwd,fwdC,rev,revC";
	string prefix = "result";
	string help="h";
	int mismatches=2;
    //Specifying the expected options
    static struct option long_options[] = {
        {"help",      no_argument,       0,  'h' },
        {"fastq1",    required_argument, 0,  '1' },
        {"fastq2",    required_argument, 0,  '2' },
        {"linker",    required_argument, 0,  'l' },
        {"prefix",    required_argument, 0,  'p' },
        {"mismatches",required_argument, 0,  'm' },
        {"direction1",required_argument, 0,  'f' },
        {"direction2",required_argument, 0,  'r' },
        {0,           0,                 0,  0   }
    };
    int opt = 0, long_index = 0;
    while ( (opt = getopt_long(argc, argv,"h1:2:l:p:m:d:f:r:", long_options, &long_index))!=-1 ){
		switch (opt) {
             case 'h' : print_usage();
			 	exit(EXIT_FAILURE);
             case '1' : fastq1 = optarg;
				 break;
             case '2' : fastq2 = optarg;
                 break;
             case 'l' : linker = optarg; 
                 break;
             case 'p' : prefix = optarg;
                 break;
             case 'm' : mismatches = atoi(optarg);
                 break;
             case 'f' : direction1 = optarg;
                 break;
             case 'r' : direction2 = optarg;
                 break;
             default: exit(EXIT_FAILURE);
        }
    }
	if ( (fastq1=="none") or (fastq2=="none") or (linker=="none") ) {
		printf("Missing one of the following options -1, -2, and/or -l.\n");
		print_usage();
		exit(EXIT_FAILURE);
	}
	// Open input and output files###########################################################
	string detailsFileName = prefix+"_details.txt";
	string mate1FileName   = prefix+"_mate1.fastq";
	string mate2FileName   = prefix+"_mate2.fastq";
	ifstream fastqFile1  (fastq1.c_str()        ); 
	ifstream fastqFile2  (fastq2.c_str()         );
	ofstream detailsFile (detailsFileName.c_str());
	ofstream mate1File   (mate1FileName.c_str()  ); 
	ofstream mate2File   (mate2FileName.c_str()  );
	// Define all possible query directions ##############################################
	map<string,string> sequences, queries1, queries2;
	sequences[ "fwd" ] = linker;                             // forward
	sequences[ "rev" ] = getReverse( sequences["fwd"]  );    // reverse
	sequences[ "fwdC" ] = getComplement( sequences["fwd"] ); // forward complement
	sequences[ "revC" ] = getComplement( sequences["rev"] ); // reverse complement
	stringstream dir1_stream(direction1), dir2_stream(direction2);
	string orientation;
	while( getline( dir1_stream,orientation,',') ) // add linker orientations specified by the user
		queries1[ orientation ] = sequences[ orientation ];
	while( getline( dir2_stream,orientation,',') ) // add liner orientations specified by the user
		queries2[ orientation ] = sequences[ orientation ];
	// Loop over fastq read mates ##########################################################
	cout << "Number of processed reads" <<endl; // Report progress
	int count = 0;
	cout << count << endl;
	string line1, line2;
	map<string,string> mate1, mate2;
	vector<string> keys;
	keys.push_back("id");keys.push_back("seq");keys.push_back("id2");keys.push_back("qual");
	while ( ! fastqFile1.eof() ) { 
		++count;
		if( count%1000000==0) cout << count << endl; // Report progress
		// Read four lines at a time
		for( unsigned i=0; i<4; ++i ){
			getline(fastqFile1,line1);	
			getline(fastqFile2,line2);
			mate1[keys[i]] = line1;
			mate2[keys[i]] = line2;	
		}
		printMates( queries1,queries2, mate1, mate2, mismatches,detailsFile,mate1File,mate2File);
	}
	// Close files #########################################################################	
	fastqFile1.close();
	fastqFile2.close();
	detailsFile.close();
	mate1File.close();
	mate2File.close();
    
	return 0;
}
